packages/contracts/src/contracts/current/forwarder/Forwarder.sol

```diff
@@ -0,0 +1,221 @@
+    /// @param feeRecipient An address of the fee recipient whom receives feeProportion of ETH.
+    /// @return Amounts filled and fees paid by maker and taker.
+    function buyTokensFee(
+        Order[] orders,
+        bytes[] signatures,
+        Order[] feeOrders,
+        bytes[] feeSignatures,
+        uint16  feeProportion,
+        address feeRecipient)
+        payable
+        public
+        returns (Exchange.FillResults memory totalFillResult)
+    {
+        require(msg.value > 0, ERROR_INVALID_INPUT);
+        address token = readAddress(orders[0].takerAssetData, 1);
+        require(token == address(ETHER_TOKEN), ERROR_INVALID_INPUT);
+        require(feeProportion <= MAX_FEE, ERROR_INVALID_INPUT);
+
+        uint256 remainingEthAmount = msg.value;
+        if (feeProportion > 0 && feeRecipient != address(0x0)) {
+            // 1.5% is 150, allowing for 2 decimal precision, i.e 0.05% is 5
+            uint256 feeRecipientFeeAmount = safeDiv(safeMul(msg.value, feeProportion), PERCENTAGE_DENOMINATOR);
+            remainingEthAmount = safeSub(msg.value, feeRecipientFeeAmount);
+            // Transfer the fee to the fee recipient
+            feeRecipient.transfer(feeRecipientFeeAmount);
+        }
+
+        ETHER_TOKEN.deposit.value(remainingEthAmount)();
+        Exchange.FillResults memory fillTokensFillResult = marketBuyTokens(orders, signatures, feeOrders, feeSignatures, remainingEthAmount);
+        addFillResults(totalFillResult, fillTokensFillResult);
+        return totalFillResult;
+    }
+
+
+    function marketBuyTokens(
+        Order[] orders,
+        bytes[] signatures,
+        Order[] feeOrders,
+        bytes[] feeSignatures,
+        uint256 sellTokenAmount)
+        private
+        returns (Exchange.FillResults memory totalFillResult)
+    {
+        uint256 takerTokenBalance = sellTokenAmount;
+        address makerTokenAddress = readAddress(orders[0].makerAssetData, 1);
+
+        Exchange.FillResults memory tokensSellQuote = marketSellOrdersQuote(orders, sellTokenAmount);
+
+        if (tokensSellQuote.takerFeePaid > 0) {
+            // Fees are required for these orders
+            // Buy enough ZRX to cover the future market sell
+            Exchange.FillResults memory feeTokensResult =
+                buyFeeTokens(feeOrders, feeSignatures, tokensSellQuote.takerFeePaid);
+            takerTokenBalance = safeSub(takerTokenBalance, feeTokensResult.takerTokenFilledAmount);
+            totalFillResult.takerFeePaid = feeTokensResult.takerFeePaid;
+        }
+
+        // Make our market sell to buy the requested tokens with the remaining balance
+        Exchange.FillResults memory requestedTokensResult = EXCHANGE.marketSellOrders(orders, takerTokenBalance, signatures);
+        // Ensure the token abstraction was fair 
+        require(isAcceptableThreshold(sellTokenAmount, requestedTokensResult.takerTokenFilledAmount), ERROR_UNACCEPTABLE_THRESHOLD);
```

Comment:
Since there might be some ZRX leftover here, the contract owner should probably have a way to withdraw the extra ZRX.

Reply:
I think that is fair, it will likely be minuscule, but if they use the forwarding contract enough it might add up. We could also deduct the current user ZRX balance from future fee abstraction and use it.